### ISSUE TITLE
Wrong dependency resolution when using dynamic versions

### ISSUE THREAD

**Issue by martofeld** [gradle/gradle#6765]
While using Gradle 4 with native dependency locking we found out an issue, given the following example we found out gradle prefers static versions over dynamic ones.

Lets say I have the following modules: 
```
1. 'com.example:module1' -> ['1.0.0']
2. 'com.example:module2' -> ['1.0.0']
3. 'com.exampe:module3' -> ['2.0.0', '2.1.0']
```

And the dependency tree is as follows
```
com.example:module1:1.0.0
\---- com.example:module3:2.+
com.example:module2:1.0.0
\---- com.example:module3:2.0.0
```

### Expected Behavior
The expected behavior would be for gradle to resolve `module3` with version `2.1.0` since is the latest 

### Current Behavior
Currently, Gradle resolves `module3` as `2.0.0` since `module2` defines it explicitly

### Context
The issue occurs while running dependencies for the module as well as locking the dependencies

### Steps to Reproduce (for bugs)
1. Use a structure similar to the one posted above
2. Run dependencies

### Your Environment
Gradle version: 4.9

**Comment by bigdaz** [comment]
Duplicate of #6102, which was fixed in `4.10`.

**Comment by bigdaz** [comment]
Actually, when dependency locking is enabled this issue persists in `4.10`. See #6340.
